cranelift/codegen/src/machinst/lower.rs

```diff
@@ -377,12 +377,33 @@ impl<'func, I: VCodeInst> Lower<'func, I> {
             }
         }
 
-        // Make a sret register, if one is needed.
+        // Find the sret register, if it's used.
         let mut sret_reg = None;
         for ret in &vcode.abi().signature().returns.clone() {
             if ret.purpose == ArgumentPurpose::StructReturn {
-                assert!(sret_reg.is_none());
-                sret_reg = Some(vregs.alloc(ret.value_type)?);
+                if let Some(entry_bb) = f.stencil.layout.entry_block() {
+                    for (i, &param) in f.dfg.block_params(entry_bb).iter().enumerate() {
+                        if vcode.abi().signature().params[i].purpose
+                            == ArgumentPurpose::StructReturn
```

Comment:
I can see that this is a pretty direct transplant of the existing logic, but would you mind making this a `zip` instead of `enumerate` plus slice indexing?